ltx/directives.tex

```diff
@@ -0,0 +1,346 @@
+\begin{itemize}
+    \item \field{locus} denotes the source location of this \grammar{empty-declaration}, which is also the source location of the sole semi-colon in the declaration
+\end{itemize}
+
+\partition{dir.empty}
+
+\note{At block scope, an \sortref{Empty}{DirSort} may be grammatically undistinguishable from an \grammar{empty-statement}.  
+    The C++ translator is expected to construct the appropriate structure based on the context of processing.}
+
+
+\subsection{\valueTag{DirSort::Attribute}}
+\label{sec:ifc:DirSort:Attribute}
+
+A \type{DirIndex} value with tag \valueTag{DirSort::Attribute} represents an abstract reference to an \grammar{attribute-declaration}.
+The \field{index} field is an index into the attribute declaration partition.
+Each entry in that partition is a structure with the following layout
+%
+\begin{figure}[H]
+    \centering
+    \structure{
+        \DeclareMember{locus}{SourceLocation} \\
+        \DeclareMember{attr}{AttrIndex} \\
+    }
+    \caption{Structure of an \grammar{attribute-declaration}}
+    \label{fig:ifc:DirSort:Attribute}
+\end{figure}
+%
+with the following meanings for the fields
+\begin{itemize} 
+    \item \field{locus} denotes the source location of this \grammar{attribute-declaration}
+    \item \field{attr} denotes the attribute (\secref{sec:ifc-attrs}) in this \grammar{attribute-declaration}
+\end{itemize}
+
+\partition{dir.attribute}
+
+\note{At block scope, an \sortref{Attribute}{DirSort} may be grammatically undistinguishable from an attributed \grammar{empty-statement}.  
+    The C++ translator is expected to construct the appropriate structure based on the context of processing.}
+
+
+\subsection{\valueTag{DirSort::Pragma}}
+\label{sec:ifc:DirSort:Pragma}
+
+A \type{DirIndex} value with tag \valueTag{DirSort::Pragma} represents an abstract reference to a pragma directive.
+The \field{index} field is an index into the pragma directive partition.
+Each entry in that partition is a structure with the following layout
+%
+\begin{figure}[H]
+    \centering
+    \structure{
+        \DeclareMember{locus}{SourceLocation} \\
+        \DeclareMember{words}{SentenceIndex} \\
+    }
+    \caption{Structure of a pragma directive}
+    \label{fig:ifc:DirSort:Pragma}
+\end{figure}
+%
+with the following meanings for the fields
+\begin{itemize}
+    \item \field{locus} denotes the source location of the entire pragma directive
+    \item \field{words} denotes the sentence index (\secref{sec:ifc-sentence}) of the words (\secref{sec:ifc-words}) making up the directive
+\end{itemize}
+
+\partition{dir.pragma}
+
+\subsection{\valueTag{DirSort::Using}}
+\label{sec:ifc:DirSort:Using}
+
+A \type{DirIndex} value with tag \valueTag{DirSort::Using} represents an abstract reference to a \grammar{using-directive} declaration.
+The \field{index} field is an index into the using-directive partition.
+Each entry in that partition is a structure with the following layout
```

Comment:
> How do we semantically extract the scope which nominated the namespace?

That would be the scope whose list of declarations this directive appears

Reply:
Does this imply you want to remove the `home_scope` field from the other declaration structures?  Based on the comment it sounds like all `home_scope` fields are redundant?

Reply:
> Does this imply you want to remove the `home_scope` field from the other declaration structures? Based on the comment it sounds like all `home_scope` fields are redundant?

No.  For a semantic declaration, the `home_scope` is relevant - and we need to distinguish them from the lexical scope.  For a directive, I am not seeing the need.  Do you have an example that would need `home_scope` (other than other semantics decls have a home scope)?

Reply:
I'm looking at this from the prospective of how quickly we can read and convey information up and down the graph.  If we go with what you suggest I:

1. Find it wildly inconsistent with other declaration types where `home_scope` is a pretty reliable field on the other structures.  Note: I'm talking specifically about `BarrenDecl`
2. Believe it makes reading the graph a bit harder as it is now the first case where I need to push down context from the scope I'm reading into the declaration resolver in order to figure out what the `home_scope` should have been.

That said, I don't believe there's a reason we _need_ to encode `home_scope` onto `BarrenDecl`, but it would be a huge convenience to authoring tooling and reduce complexity in the compiler as I would need to add more context to inner resolvers whereas before I could pass the structure `X` directly and have all the information I need to resolve enclosing scopes.

This is mostly a balance between ideal representation (which is what you have here) and duplicate, but useful, information.

Reply:
We are talking about directives here, not semantic declarations for which `home_scope` is an essential component.   Directives always appear on the list of something, even though they don't _belong_ to a thing.

I would like for us to proceed with the minimal representation here for the directives and come up with evidence about how inconvenient it is and add the extra info - it is harder to remove than to add.